x/streamer/keeper/pump_stream.go

```diff
@@ -0,0 +1,370 @@
+// randomInRange = randomBig % 1000 = 890
+//
+// Since we want 25 pumps out of 1000 blocks:
+// - Numbers 0-24 → pump (25 numbers)
+// - Numbers 25-999 → no pump (975 numbers)
+//
+// 890 ≥ 25, so no pump this block
+func ShouldPump(
+	ctx sdk.Context,
+	budget math.Int,
+	left math.Int,
+	pumpNum uint64,
+	epochBlocks math.Int, // is consumed!
+) (math.Int, error) {
+	if pumpNum == 0 {
+		// Should not pump at all
+		return math.ZeroInt(), nil
+	}
+	if epochBlocks.IsZero() {
+		return math.ZeroInt(), fmt.Errorf("epochBlocks cannot be zero")
+	}
+	if pumpNum > epochBlocks.Uint64() {
+		return math.ZeroInt(), fmt.Errorf("pumpNum (%d) cannot be greater than epochBlocks (%d)", pumpNum, epochBlocks)
+	}
+
+	// Scale down the random value to range [0, epochBlocks)
+	randomInRange := GenerateUnifiedRandom(ctx, epochBlocks.BigIntMut())
+
+	// Check if the random value falls within the pump probability
+	// For pumpNum pumps in epochBlocks: success if random value < pumpNum
+	numeratorBig := big.NewInt(int64(pumpNum))
+
+	// If randomInRange < numeratorBig
+	if randomInRange.Cmp(numeratorBig) < 0 {
+		return PumpAmt(ctx, budget, left, pumpNum)
+	}
+
+	return math.ZeroInt(), nil
+}
+
+// PumpAmt computes min(Uniform(0, 2 * Budget / PumpNum), Left).
+func PumpAmt(ctx sdk.Context, budget math.Int, left math.Int, pumpNum uint64) (math.Int, error) {
+	modulo := budget.MulRaw(2).QuoRaw(int64(pumpNum))
+	if modulo.IsZero() {
+		return math.ZeroInt(), fmt.Errorf("budget per pump is fractional: too small budget (%s) or too many pumps (%d)", budget, pumpNum)
+	}
+	randBig := GenerateUnifiedRandom(ctx, modulo.BigIntMut())
+	rand := math.NewIntFromBigIntMut(randBig)
+	return math.MinInt(rand, left), nil
+}
+
+// GenerateUnifiedRandom a unified random variable by modulo.
+func GenerateUnifiedRandom(ctx sdk.Context, modulo *big.Int) *big.Int {
+	h := sha256.New()
+	h.Write(ctx.HeaderHash())
+	// h.Write(blockTimeBytes)
+	seed := h.Sum(nil)
+
+	randomBig := new(big.Int).SetBytes(seed)
+	return new(big.Int).Mod(randomBig, modulo)
+}
+
+// ExecutePump performs the pump operation by buying tokens for a specific rollapp.
+// CONTRACT: pumpAmt is always in base denom.
+func (k Keeper) ExecutePump(
+	ctx sdk.Context,
+	pumpAmt sdk.Coin,
+	rollappID string,
+) (tokenOut sdk.Coin, err error) {
+	rollapp, found := k.rollappKeeper.GetRollapp(ctx, rollappID)
+	if !found {
+		return sdk.Coin{}, fmt.Errorf("rollapp not found: %s", rollappID)
+	}
+
+	plan, found := k.iroKeeper.GetPlanByRollapp(ctx, rollapp.RollappId)
+	if !found {
+		return sdk.Coin{}, fmt.Errorf("IRO plan not found for rollapp: %s", rollappID)
+	}
+
+	buyer := k.ak.GetModuleAddress(types.ModuleName)
+
+	var tokenOutDenom string
+	if plan.IsSettled() {
+		tokenOutDenom = plan.SettledDenom
+	} else {
+		tokenOutDenom = plan.LiquidityDenom
+	}
+	var tokenOutAmt math.Int
+
+	err = osmoutils.ApplyFuncIfNoError(ctx, func(ctx sdk.Context) error {
+		// There are several cases:
+		// - targetDenom == plan.SettledDenom =>
```

Comment:
plan.SettledDenom means the denom of the rollapp after the plan has been settled. this is what you mean here? I think not. I believe you mean the rollap denom pair for the pool (i.e DYM or USDC)?